File: 2024/day14.go
package main

import (
	"bytes"

	"lukechampine.com/advent/utils"
)

var input = utils.Input(2024, 14)

const maxX = 101
const maxY = 103

type robot struct {
	pos utils.Pos
	vel utils.Pos
}

func updateRobots(robots []robot) {
	for i := range robots {
		r := &robots[i]
		r.pos = r.pos.Add(r.vel.X, r.vel.Y)
		r.pos.X = (r.pos.X + maxX) % maxX
		r.pos.Y = (r.pos.Y + maxY) % maxY
	}
}

func renderRobots(robots []robot) [][]byte {
	grid := utils.ByteGrid(maxX, maxY, '.')
	for _, r := range robots {
		grid[r.pos.Y][r.pos.X] = '#'
	}
	return grid
}

func gridEntropy(grid [][]byte) (runs int) {
	flat := bytes.Join(grid, nil)
	for i := range flat[1:] {
		runs += utils.BoolToInt(flat[i] != flat[i+1])
	}
	return
}

func main() {
	var robots []robot
	for _, line := range utils.Lines(input) {
		ints := utils.ExtractInts(line)
		robots = append(robots, robot{
			pos: utils.Pos{X: ints[0], Y: ints[1]},
			vel: utils.Pos{X: ints[2], Y: ints[3]},
		})
	}

	for i := 0; i < 100; i++ {
		updateRobots(robots)
	}
	var quads [4]int
	for i, min := range []utils.Pos{
		{X: 0, Y: 0},
		{X: maxX/2 + 1, Y: 0},
		{X: 0, Y: maxY/2 + 1},
		{X: maxX/2 + 1, Y: maxY/2 + 1},
	} {
		quad := utils.Rect{Min: min, Max: min.Add(maxX/2-1, maxY/2-1)}
		for _, r := range robots {
			quads[i] += utils.BoolToInt(quad.Contains(r.pos))
		}
	}
	utils.Println(quads[0] * quads[1] * quads[2] * quads[3])

	// minimize entropy
	minE := gridEntropy(renderRobots(robots))
	longevity := 0
	for seconds := 100; ; seconds++ {
		if e := gridEntropy(renderRobots(robots)); e < minE {
			minE = e
			longevity = 0
		} else if longevity++; longevity == 10000 {
			utils.Println(seconds - 10000)
			break
		}
		updateRobots(robots)
	}
}

File: 2024/rust/src/day14.rs
use crate::utils;

const MAX_X: isize = 101;
const MAX_Y: isize = 103;

struct Robot {
    pos: utils::Point,
    vel: utils::Point,
}

fn mod_like_go(a: isize, b: isize) -> isize {
    ((a % b) + b) % b
}

impl Robot {
    fn from_string(s: &str) -> Self {
        let ints = utils::ints(s).collect::<Vec<_>>();
        Robot {
            pos: (ints[0], ints[1]).into(),
            vel: (ints[2], ints[3]).into(),
        }
    }

    fn update(&mut self) {
        self.pos = self.pos.add(self.vel);
        self.pos.x = mod_like_go(self.pos.x + MAX_X, MAX_X);
        self.pos.y = mod_like_go(self.pos.y + MAX_Y, MAX_Y);
    }
}

pub fn part1(input: &str) -> String {
    let mut robots = input
        .lines()
        .map(|line| Robot::from_string(line))
        .collect::<Vec<_>>();
    for _ in 0..100 {
        for r in &mut robots {
            r.update();
        }
    }
    vec![
        utils::Point::from((0, 0)),
        utils::Point::from((MAX_X / 2 + 1, 0)),
        utils::Point::from((0, MAX_Y / 2 + 1)),
        utils::Point::from((MAX_X / 2 + 1, MAX_Y / 2 + 1)),
    ]
    .iter()
    .map(|min| {
        robots
            .iter()
            .filter(|r| {
                min.x <= r.pos.x
                    && r.pos.x <= min.x + MAX_X / 2 - 1
                    && min.y <= r.pos.y
                    && r.pos.y <= min.y + MAX_Y / 2 - 1
            })
            .count()
    })
    .product::<usize>()
    .to_string()
}

pub fn part2(input: &str) -> String {
    let mut robots = input
        .lines()
        .map(|line| Robot::from_string(line))
        .collect::<Vec<_>>();

    let mut grid = vec![false; (MAX_X * MAX_Y) as usize];
    std::iter::from_fn(move || {
        grid.fill(false);
        for r in &mut robots {
            r.update();
            grid[(r.pos.y * MAX_X + r.pos.x) as usize] = true;
        }
        Some(grid.windows(2).filter(|w| w[0] != w[1]).count())
    })
    .enumerate()
    .take(10000)
    .min_by_key(|x| x.1)
    .unwrap()
    .0
    .to_string()
}
